Integrate Vetiver, Quarto, and refactor

SamEdwardes opened the pull request:
This PR is a major update to the bikeshare example. Closes https://github.com/rstudio/sol-eng/issues/114. Changes include:

- Refactoring and reorganizing the code
- Integration of vetiver

## Refactoring and reorganizing the code

I have refactored the app to be more organized. Major changes include:

- All deployed content now lives in the `content` directory
- Content is numbered to reflect the order that it should be run in (where applicable).
- A new script `deploy_all_content.qmd` was created. This script is meant to be run from Workbench and will deploy all of the content to Connect with the correct titles, tags, etc.

## Integration of vetiver

The model is now deployed to connect using vetiver. When making API calls vetiver is also used.

SamEdwardes commented:
@gsingh91 and @xuf12 this PR is ready for review.

gsingh91 commented on content/02-model/01-train-and-deploy-model.qmd:
```diff
@@ -0,0 +1,300 @@
+content$update(access_type = "all")
+
+# Define the contents collaborators
+collaborators <- c(
```
We should remove @xuf12 and me from auto collaborators since we are publishing this repo publicly.

SamEdwardes replied:
I wanted to include you as collaborators so that you can make changes to the content if needed. E.g. I am on vacation and something goes wrong with the content. If I am the only collaborator only I can make changes.

What do you think?

SamEdwardes replied:
I removed all of the collaborator logic in 618731f.

gsingh91 commented on content/02-model/01-train-and-deploy-model.qmd:
````diff
@@ -0,0 +1,300 @@
+readLines("api/plumber.R") %>%
+  paste(collapse = "\n") %>%
+  cat()
+```
````
is there a reason the `plumber.R` file is not included? We should have one version of how this file looks like in the repo.

SamEdwardes replied:
I chose not to include `plumber.R` because it is an auto generated bit of code and should not be edited by hand. I do print the contents of `plumber.R` just to demonstrate to readers what the output of `vetiver::vetiver_write_plumber` is.

SamEdwardes replied:
Updated code to keep api/plumber.R in source code: d398a51.

gsingh91 commented on content/02-model/02-model-metrics.qmd:
````diff
@@ -0,0 +1,149 @@
+
+```{r disconnect_db}
+odbc::dbDisconnect(con)
+```
````
this file would also need rework when metrics monitoring is available in `vetiver`

SamEdwardes replied:
Agreed - no actions for the time being though I think.

SamEdwardes commented:
Updates finished. Ready for review next review @gsingh91.

akgold commented on content/01-etl/01-raw-data-refresh.qmd:
```diff
@@ -0,0 +1,148 @@
+library(dplyr)
+library(odbc)
+
+# The following packages are not directly called, but are requirements. They
```
I *think* these can probably go away now that I fixed the `bikeHelpR` dependencies. You'll have to reinstall and update the manifest.

SamEdwardes replied:
Updated, see 85cd664.

akgold reviewed (approved):
Few little things in code -- it might be worthwhile to push changes to `bikeHelpR` first so you can pick them up and adjust code.

My one overarching comment is that I'm not sure a deploy script is a good idea for this as a demo asset. Using git-backed deployment was one of the key features here to demo. See image below 
<img width="681" alt="image" src="https://user-images.githubusercontent.com/394930/169828490-02a8c256-ad73-449e-901e-6babe55c715e.png">

So using a deploy script is more convenient...but might be a bad choice from the perspective of using this as a demo app. I think this also applies to the choice of whether to include the `plumber.R` in the repo (I'd argue yes).

SamEdwardes commented:
Thanks for the review @akgold, I will work on the updates. 

One thing that I was thinking, should we move BikeHelpR to its own repo so that it is easier to maintain and then we can point Colorado RSPM (and public package manager) to that repo instead of a subdirectory in this repo?

It could live here: https://github.com/sol-eng/bikeHelpR

SamEdwardes commented on content/01-etl/02-tidy-data.qmd:
```diff
@@ -0,0 +1,73 @@
+# The following packages are not directly called, but are requirements. They
+# need to be here so that `rsconnect::writeManifest` is able to capture them
+# as a dependency.
+library(dbplyr)
```
Updated, see 85cd664.

SamEdwardes commented:
PR is ready to be merged. The following updates were made:

- Deleted the script to deploy all apps, instead git based deploys will be set up after merging.
- Added a script to quickly edit and update deployed content metadata.
- Moved bikeHelpR to its own repo

jthomasmock commented:
Howdy @SamEdwardes - while I'm very excited about Quarto + bikeshare, I don't think the current content items came across as a Quarto files.

I saw that the files still "looked" like RMarkdown as opposed to the new styling in Quarto, so I pulled the [content details](https://docs.rstudio.com/connect/api/#get-/v1/experimental/content/{guid}) for the [ETL Step 1](https://colorado.rstudio.com/rsc/connect/#/apps/e1e47cc0-5c31-41bd-a1f4-5729b2ece52e/access/5382)

```
library(httr)

apiKey <- Sys.getenv("CONNECT_API_KEY")

result <- GET(glue::glue("{Sys.getenv('CONNECT_SERVER')}/__api__/v1/experimental/content/e1e47cc0-5c31-41bd-a1f4-5729b2ece52e"),
              add_headers(Authorization = paste("Key", apiKey)))
output <- content(result)

# came across as rmd-static instead of quarto-static
output$title
#> [1] "Bike Predict - ETL Step 1 - Raw Data Refresh"
output$app_mode
#> [1] "rmd-static"
# no quarto used in bundle
output$quarto_version
#> NULL
```

It should have a `quarto_version` printed ie:

```
r_version: "3.5.1",
py_version: "3.8.2",
quarto_version: "0.2.22", # shouldn't be NULL
run_as: "rstudio-connect",
run_as_current_user: false,
owner_guid: "25438b83-ea6d-4839-ae8e-53c52ac5f9ce",
url: "http://rstudio-connect.company.com/content/42",
role: "owner"
},
```

---

### Potential fix

To resolve, you will probably need to use something like:

```
# quarto R package has a helper function for quarto_path, but could
# report it manually from terminal w/ `$ which quarto`
rsconnect::writeManifest(appPrimaryDoc = "2022-05-26.qmd", quarto = quarto::quarto_path())
```

Which should generate a manifest file that contains Quarto:

```
{
  "version": 1,
  "locale": "en_US",
  "platform": "4.2.0",
  "metadata": {
    "appmode": "quarto-static",
    "primary_rmd": "2022-05-26.qmd",
    "primary_html": null,
    "content_category": null,
    "has_parameters": false
  },
  "quarto": {
    "version": "0.9.387",
    "engines": ["knitr"]
  },
  "packages": {
```

as opposed to the current manifest file:

```
{
  "version": 1,
  "locale": "en_US",
  "platform": "4.1.2",
  "metadata": {
    "appmode": "rmd-static",
    "primary_rmd": "document.qmd",
    "primary_html": null,
    "content_category": null,
    "has_parameters": false
  },
  "packages": {
```

### Quarto + Colorado

All that being said, I still don't think that Colorado is ready for Quarto items as the RSC 2022.05.0 release had a bug fix for Quarto + K8S on RSC. So it may be fine to "leave as is" given that Quarto content won't be publishable until Cole updates the Colorado RSC server to 2022.05.0.

I'm getting errors like the below still whether deploying from `rsconnect` or push-button, 

note the specific to K8S errors that was fixed in 2022.05.0:
- `whoami: cannot find name for user ID 999`  
- `Warning message: In system("whoami", intern = TRUE) : running command 'whoami' had status 1`

```
rsconnect::deployDoc("2022-05-26.qmd", quarto = quarto::quarto_path(), server = "colorado.rstudio.com")
Discovering document dependencies... OK
Preparing to deploy document...DONE
Uploading bundle for document: 11476...DONE
Deploying bundle: 56960 for document: 11476 ...
[Connect] Building Quarto document...
[Connect] Bundle created with R version 4.2.0 and Quarto version 0.9.387 is compatible with environment Kubernetes::ghcr.io/rstudio/content-pro:r4.1.3-py3.10.4-bionic with R version 4.1.3 from /opt/R/4.1.3/bin/R and Quarto version 0.9.344 from /opt/quarto/bin/quarto 
[Connect] Bundle requested R version 4.2.0; using /opt/R/4.1.3/bin/R from Kubernetes::ghcr.io/rstudio/content-pro:r4.1.3-py3.10.4-bionic which has version 4.1.3
[Connect] Determining session server location ...
[Connect] Connecting to session server http://172.19.16.23:30285 ...
[Connect] Connected to session server http://172.19.16.23:30285
[Connect] 2022/05/27 03:02:42.967646992 whoami: cannot find name for user ID 999
[Connect] 2022/05/27 03:02:42.969423329 Warning message:
[Connect] 2022/05/27 03:02:42.969431955 In system("whoami", intern = TRUE) : running command 'whoami' had status 1
.... MORE STUFF

.... THEN
[Connect] Using environment Kubernetes::ghcr.io/rstudio/content-pro:r4.1.3-py3.10.4-bionic
[Connect] Using /opt/quarto/bin/quarto with version 0.9.344 from Kubernetes::ghcr.io/rstudio/content-pro:r4.1.3-py3.10.4-bionic
[Connect] Using /opt/R/4.1.3/bin/R with version 4.1.3 from Kubernetes::ghcr.io/rstudio/content-pro:r4.1.3-py3.10.4-bionic
[Connect] Determining session server location ...
[Connect] Connecting to session server http://172.19.16.23:32751 ...
[Connect] 2022/05/27 03:05:54.849767132 [rsc-quarto] Error: cannot lookup current user: user: unknown userid 999
[Connect] Unable to render the deployed content: Rendering exited abnormally: nonzero exit status: 1
[Connect] Stopped connection attempts to session server http://172.19.16.23:32751
Document deployment failed with error: Rendering exited abnormally: nonzero exit status: 1
```

SamEdwardes commented:
@jthomasmock thank for the very helpful right up here! Indeed your initial guess was right, we are not publishing quarto. I incorrect assumed that since it was a .qmd file that writeManifest was correctly determining it to be a quarto doc. I did not inspect manifest.json to see that it was actually static rmd!

```json
{
  "version": 1,
  "locale": "en_US",
  "platform": "4.1.2",
  "metadata": {
    "appmode": "rmd-static",
    "primary_rmd": "document.qmd",
    "primary_html": null,
    "content_category": null,
    "has_parameters": false
  },
  "packages": {
```

## Colorado

I double checked, and I came across the same errors as you when attempting a git backed deploy to Colorado.

```
2022/05/27 13:27:27.247818736 [rsc-quarto] Error: cannot lookup current user: user: unknown userid 999
```

So I agree with you suggestion we should leave as is until we get the 2022.05.0 update.
